src/coreComponents/physicsSolvers/fluidFlow/ReactiveCompositionalMultiphaseOBLKernels.hpp

```diff
@@ -0,0 +1,1427 @@
+  ElementViewConst< arrayView2d< real64 const, compflow::USD_COMP > > const m_compFrac;
+  ElementViewConst< arrayView2d< real64 const, compflow::USD_COMP > > const m_dCompFrac;
+
+  /// Views on temperature
+  ElementViewConst< arrayView1d< real64 const > > const m_temp;
+  ElementViewConst< arrayView1d< real64 const > > const m_dTemp;
+
+
+  // Views on OBL operators and their derivatives
+  ElementViewConst< arrayView2d< real64 const, compflow::USD_OBL_VAL > > const m_OBLOperatorValues;
+  ElementViewConst< arrayView3d< real64 const, compflow::USD_OBL_DER > > const m_OBLOperatorDerivatives;
+
+  // Residual and jacobian
+
+  /// View on the local CRS matrix
+  CRSMatrixView< real64, globalIndex const > const m_localMatrix;
+  /// View on the local RHS
+  arrayView1d< real64 > const m_localRhs;
+};
+
+/**
+ * @class FaceBasedAssemblyKernel
+ * @tparam NUM_PHASES number of phases
+ * @tparam NUM_COMPS number of components
+ * @tparam ENABLE_ENERGY flag if energy balance equation is assembled
+ * @tparam STENCILWRAPPER wrapper for element stencils
+ * @brief Compute flux term for an element
+ */
+template< integer NUM_PHASES, integer NUM_COMPS, bool ENABLE_ENERGY, typename STENCILWRAPPER >
+class FaceBasedAssemblyKernel : public FaceBasedAssemblyKernelBase
+{
+public:
+
+  /// Compile time value for the number of phases
+  static constexpr integer numPhases = NUM_PHASES;
+  /// Compile time value for the number of components
+  static constexpr integer numComps = NUM_COMPS;
+  /// Compile time value for the energy balance switch
+  static constexpr integer enableEnergyBalance = ENABLE_ENERGY;
+
+  /// Compile time value for the number of primary variables
+  static constexpr integer numDofs = numComps + enableEnergyBalance;
+
+  /// Compile time value for the number of equations
+  static constexpr integer numEqns = numDofs;
+
+  // /// Compile time value for the number of operators
+  static constexpr integer numOps = COMPUTE_NUM_OPS( NUM_PHASES, NUM_COMPS, ENABLE_ENERGY );
+
+  // component flux
+  static constexpr integer FLUX_OP = numDofs;
+  // diffusion
+  static constexpr integer UPSAT_OP = numDofs + numDofs * numPhases;
+  static constexpr integer GRAD_OP = numDofs + numDofs * numPhases + numPhases;
+
+  // temperature
+  static constexpr integer RE_TEMP_OP = numDofs + numDofs * numPhases + numPhases + numDofs * numPhases + numDofs + 1;
+  // rock conduction
+  static constexpr integer ROCK_COND = numDofs + numDofs * numPhases + numPhases + numDofs * numPhases + numDofs + 2;
+  // gravity (density)
+  static constexpr integer GRAV_OP = numDofs + numDofs * numPhases + numPhases + numDofs * numPhases + numDofs + 3;
+  // capillary pressure
+  static constexpr integer PC_OP = numDofs + numDofs * numPhases + numPhases + numDofs * numPhases + numDofs + 3 + numPhases;
+  // porosity
+  static constexpr integer PORO_OP = numDofs + numDofs * numPhases + numPhases + numDofs * numPhases + numDofs + 3 + 2 * numPhases;
```

Comment:
Is there a reason why these do not follow a consistent naming convention?
I expect most of these names to start with `num`?

Reply:
I agree this is a little confusing: these `integer` are in fact not quantities (`numX`) but offsets used to accessed the bundled operator values (+ derivatives) in the arrays `OBLOperatorValues` and `OBLOperatorValues_n`. 

So, maybe a good naming convention would be `FLUX_OP_OFFSET` for instance? What do you think?